Sources/StreamChat/Utils/MessagesPaginationStateHandling/MessagesPaginationThreadSafeDecorator.swift

```diff
@@ -0,0 +1,33 @@
+//
+// Copyright © 2023 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+/// A decorator of the `MessagesPaginationStateHandling` to make sure it is thread safe.
+class MessagesPaginationThreadSafeDecorator: MessagesPaginationStateHandling {
```

Comment:
Why is this one needed rather than making `MessagesPaginationStateHandler` thread-safe?

Reply:
This approach makes sure that the Threading logic is separate from the actual business logic. It makes the core component easier to read and also easier to test. In our codebase, we usually put this responsibility inside the actual implementation, but this time, because this component has a lot of logic, I used a decorator to make the logic easy to read. As the component grows, the threading logic will also grow, and probably will also add logging as well, so the Decorator pattern is really useful to decouple Threading, Logging, etc... from the main business logic.

Reply:
I kind of agree with Ilias here, I think it's better that the component is thread-safe by default, instead of decorating it. Few arguments for that:
- someone can make the mistake of using the thread unsafe state directly.
- if the thread safe implementation is an internal detail, we can easily change it with e.g. actors in the future.
- we would actually find it harder to maintain it, since for every new addition in the state, we would need to expose sth in the decorator.
- consistency with the rest of the codebase.
- also, not sure how would this help in testing, can you clarify?

Reply:
> someone can make the mistake of using the thread unsafe state directly.

I think that this is very unlikely, since the developer usually we check how it is used in other places

> if the thread-safe implementation is an internal detail, we can easily change it with e.g. actors in the future.

The goal of this decorator pattern is to actually have an "actor" feel in terms of usability.

> we would actually find it harder to maintain it, since, for every new addition in the state, we would need to expose sth in the decorator.

This is actually a good thing. It makes sure you don't forget to handle new functions that you add.

> consistency with the rest of the codebase

This one is valid, but there are multiple places in our codebase where we don't use this pattern, and 100% should. The most obvious one is the Updaters, where we want to make sure all the completions are executed in the `callbackQueue`. With this pattern it would be much cleaner, and also less error-prone.

> also, not sure how would this help in testing, can you clarify

In this particular case, it won't help because we always use `sync` instead of `async` but when using `async` it makes the test harder because you need to mock the queue.

Maybe for this scenario, it is not a clear win using this, not as much as the Updaters, and since you guys don't agree, I'll make the component thread-safe directly.

Reply:
@ipavlidakis @martinmitrevski I've removed it here: https://github.com/GetStream/stream-chat-swift/pull/2595/commits/f254307e3b8daa1f7ce5cd04033aac07db0e9286, let me know what you guys think.

Keep in mind it is not a good idea to do what we usually do, something like this:
```swift
var state: MessagesPaginationState {
        get {
            queue.sync {
                _state
            }
        }
        set {
            queue.sync {
                _state = newValue
            }
        }
    }
```
Because we change **multiple** properties of the `state` at the same time, which would trigger a `queue.sync` multiple times.